Invalid "a constant with a similar name exists: `_`" suggestion when a `const _` is in scope

For example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=const%20_%3A%20()%20%3D%20()%3B%0A%0Afn%20main()%20%7B%0A%20%20%20%20a%0A%7D)):

```rust
const _: () = ();

fn main() {
    a
}
```

Errors with:

```
error[E0425]: cannot find value `a` in this scope
 --> src/main.rs:4:5
  |
1 | const _: () = ();
  | ----------------- similarly named constant `_` defined here
...
4 |     a
  |     ^ help: a constant with a similar name exists: `_`
```

The same also happens with `use Trait as _` ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&code=pub%20use%20Iterator%20as%20_%3B%0A%0Afn%20foo%3CT%3A%20A%3E(x%3A%20T)%20%7B%7D)):

```rust
pub use Iterator as _;

fn foo<T: A>(x: T) {}
```

```
error[E0405]: cannot find trait `A` in this scope
    --> src/lib.rs:3:11
     |
3    |   fn foo<T: A>(x: T) {}
     |             ^ help: a trait with a similar name exists: `_`
```

The diagnostics shouldn't suggest things named underscore.

--- a/src/librustc_resolve/diagnostics.rs
+++ b/src/librustc_resolve/diagnostics.rs
@@ -769,6 +769,11 @@ impl<'a> Resolver<'a> {
         span: Span,
     ) -> bool {
         if let Some(suggestion) = suggestion {
+            // We shouldn't suggest underscore.
+            if suggestion.candidate == kw::Underscore {
+                return false;
+            }
+
             let msg = format!(
                 "{} {} with a similar name exists",
                 suggestion.res.article(),

--- a/src/test/ui/resolve/typo-suggestion-named-underscore.rs
+++ b/src/test/ui/resolve/typo-suggestion-named-underscore.rs
@@ -1,0 +1,14 @@
+const _: () = ();
+
+fn main() {
+    a // Shouldn't suggest underscore
+    //~^ ERROR: cannot find value `a` in this scope
+}
+
+trait Unknown {}
+
+#[allow(unused_imports)]
+use Unknown as _;
+
+fn foo<T: A>(x: T) {} // Shouldn't suggest underscore
+//~^ ERROR: cannot find trait `A` in this scope

--- a/src/test/ui/resolve/typo-suggestion-named-underscore.stderr
+++ b/src/test/ui/resolve/typo-suggestion-named-underscore.stderr
@@ -1,0 +1,16 @@
+error[E0425]: cannot find value `a` in this scope
+  --> $DIR/typo-suggestion-named-underscore.rs:4:5
+   |
+LL |     a // Shouldn't suggest underscore
+   |     ^ not found in this scope
+
+error[E0405]: cannot find trait `A` in this scope
+  --> $DIR/typo-suggestion-named-underscore.rs:13:11
+   |
+LL | fn foo<T: A>(x: T) {} // Shouldn't suggest underscore
+   |           ^ not found in this scope
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0405, E0425.
+For more information about an error, try `rustc --explain E0405`.

